File: src/auto_control.cpp
#include <iostream>
#include <ros/ros.h>
#include <kkctbn2019/AutoControl.h>
#include <mavros_msgs/RCIn.h>

ros::Publisher auto_control_publisher;


void pwmModeCallback(const mavros_msgs::RCIn::ConstPtr& msg) {
    int pwm = msg->channels[6];
    kkctbn2019::AutoControl control;
    if (pwm < 1400) {
        control.state = kkctbn2019::AutoControl::AVOID_RED;
    } else if (pwm > 1600) {
        control.state = kkctbn2019::AutoControl::AVOID_GREEN;
    } else {
        control.state = kkctbn2019::AutoControl::AVOID_RED_AND_GREEN;
    }
    mode_publisher.publish(mode);
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "mode");
    ros::NodeHandle nh;

    ros::Subscriber pwm_mode_subscriber = nh.subscribe("/mavros/rc/in", 8, pwmModeCallback);

    auto_control_publisher = nh.advertise<kkctbn2019::AutoControl>("/makarax/auto_control", 8);

    ROS_WARN("mode is active");

    ros::spin();
    
    return 0;
}

File: scripts/auto_control.py
#!/usr/bin/env python
import rospy
from mavros_msgs.msg import RCIn
from kkctbn2019.msg import AutoControl

def rc_callback(msg):
    try:
        pwm = msg.channels[6];
    except IndexError:
        pwm = 1300

    control = AutoControl()
    if pwm < 1400:
        control.state = AutoControl.MANUAL;
    elif pwm > 1600:
        control.state = AutoControl.AVOID_RED_AND_GREEN;
    else:
        control.state = AutoControl.AVOID_RED;

    auto_control_publisher.publish(control)

if __name__  == '__main__':
    rospy.init_node("auto_control")

    rc_subscriber = rospy.Subscriber("/mavros/rc/in", RCIn, rc_callback)
    auto_control_publisher = rospy.Publisher("/makarax/auto_control", AutoControl, queue_size=8)
    
    rospy.spin()
